### PR TITLE
MAINTAINING.md: Added doc on release process

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Added informations relevant to maintainers on how the normal review process differs during the feature freeze, or when the are targeting a release rather than the master branch.

### Testing procedure

Doesn't apply

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/pull/12486

### CODE REVIEW

File: MAINTAINING.md [hunk 1]
```diff
@@ -155,6 +155,25 @@ GitHub's review sidebar.
...
+freeze only PRs with minor impact should be merged into master. The hard feature
+freeze begins when the release manager creates a new release branch. Therefore,
+the restriction on merging PRs into the master branch are lifted at that point.
+
+Once the release branch is created, no new features should be backported.
+Instead, backports should consist only of bug fixes or of reverting features
+that were added during the last development period (and not part of any
+release), but didn't reach the required maturity or API stability yet. For
+bigger changes (which explicitly includes any revert), the PR has to be
+announced in the maintainer mailing list and should be merged no sooner than
```
Comment (miri64) [inline MAINTAINING.md]:
```suggestion
announced to the maintainer mailing list and should be merged no sooner than
```

File: MAINTAINING.md [hunk 2]
```diff
@@ -155,6 +155,25 @@ GitHub's review sidebar.
...
+48h after the announcement and needs at least to ACKs. In case of security
+relevant fixes, the PR can be merged 1h after the announcement.
```
Comment (miri64) [inline MAINTAINING.md]: Do you mean just fixes or also reverts?

**Comment by miri64** [comment]
Don't want it to be appear that I'm the only one responsible for this update to the maintainer guidelines, so I removed the assignment to me.

File: MAINTAINING.md (hunk 2 above)
Comment (maribu) [reply MAINTAINING.md]: In the security relevant part? Than yes, I think tying the hands of someone waiting to hit the merge button on a critical security fix is a bad idea.

In the announcement part? I'd say only "high impact" bug fixes should be announced.

Comment (miri64) [reply MAINTAINING.md]: > In the security relevant part? Than yes, I think tying the hands of someone waiting to hit the merge button on a critical security fix is a bad idea.

I meant the security relevant part. Should security relevant reverts (e.g. a feature broken beyond repair) also be under that rule. If yes, please adopt the sentence so it is clearer.

> In the announcement part? I'd say only "high impact" bug fixes should be announced.

I think security bugs don't need to be announced to maintainers, unless they are so big it should involve some more pairs of eyes than just two (but I guess that is what you are saying).

Comment (maribu) [reply MAINTAINING.md]: I think I adapted the wording to be clearer. Please hit "unresolve" if you disagree.

**Comment by maribu** [comment]
I think a follow up PR would be reasonable to also document other special handling during review, as:

- Changes in core need two ACKs
- API changes need two ACKs
    - Deprecation process is needed on API changes if the API was included in a release so far
- ...

But I guess it is better to keep this PR focused.

**Comment by miri64** [comment]
> API changes need two ACKs
> 
>     * Deprecation process is needed on API changes if the API was included in a release so far

I think we should also discuss this. I heard this now a few times by different people, but I can't remember that we ever decided on that.

File: MAINTAINING.md [hunk 3]
```diff
@@ -155,6 +155,28 @@ GitHub's review sidebar.
...
+
+Before the official release of a new RIOT version, two feature freeze periods
+are announced on the
+[RIOT development email list](https://lists.riot-os.org/mailman/listinfo/devel):
+The soft feature freeze and the hard feature freeze. During the soft feature
+freeze only PRs with minor impact should be merged into master. The hard feature
+freeze begins when the release manager creates a new release branch. Therefore,
+the restriction on merging PRs into the master branch are lifted at that point.
+
+Once the release branch is created, no new features should be backported.
```
Comment (maribu) [inline MAINTAINING.md]: I would now prefer to reword it to "Once the release branch is created, no backports of new features should be accepted". I think that we cannot expect e.g. a new contributor to be aware of the release process. The original sentence would however blame that person, if she/he creates a PR to backport a feature.

I think however it is fair to expect that maintainers are aware of `MAINTAINING.md` and its contents.

File: MAINTAINING.md [hunk 4]
```diff
@@ -155,6 +155,28 @@ GitHub's review sidebar.
...
+Instead, backports should consist only of bug fixes or of reverting features
+that were added during the last development period (and not part of any
+release), but didn't reach the required maturity or API stability yet. For
+bigger changes (which explicitly includes any revert), the PR has to be
+announced to the maintainer mailing list and should be merged no sooner than
+48h after the announcement and needs at least to ACKs.
```
Comment (maribu) [inline MAINTAINING.md]: Here the formulation to me means that the contributor opening the PR containing a backported bugfix of high impact is not required to post it to the maintainer mailing list.(And also no maintainer reviewing it is explicitly required to do so. It is only required that no one hits the merge button before the PR has been announced and the 48h have passed.

To me this makes a lot of sense: Again, the contributor should not be required to know the release process. Also, a maintainer might to a partial review - maybe even only adding label to categorizes - and does not realize that the PR could be considered as having a high impact.

But I think it is fair to expect maintainers to not hit the merge button on PRs targeting a release branch if the cannot reasonably decide themselves if a PR is a high impact PR or not.

Comment (benpicco) [reply MAINTAINING.md]: I think if a new feature turned out to cause a regression in existing code, it should be reverted without much discussion.
The feature is then postponed to the next release if the bug has been fixed, but in a hard freeze you don't want to have 'scary' things.

File: MAINTAINING.md (hunk 3 above)
Comment (fjmolinas) [reply MAINTAINING.md]: I would like to rephrase to `"Once the release branch is created, no backports of new features will be accepted"`.

File: MAINTAINING.md (hunk 4 above)
Comment (fjmolinas) [reply MAINTAINING.md]: > I think if a new feature turned out to cause a regression in existing code, it should be reverted without much discussion.

For me this depends if the regression is hard/easy to fix, it would depend on contributors reactivity. But ultimately this is a Release manager decision, discussion can happen but if the manger considers it 'scary', the I agree there should not be much discussion.

File: MAINTAINING.md (hunk 3 above)
Comment (maribu) [reply MAINTAINING.md]: @fjmolinas: Added, with slightly different wording. Can you confirm you're fine with that?

**Comment by fjmolinas** [comment]
I think we pretty much agree on this one right @miri64 @maribu? IMO this is ok to squash.

**Comment by maribu** [comment]
Squashed

**Comment by aabadie** [comment]
The 2 required ACKs are there on this one. @miri64 @fjmolinas do they still hold ?

**Comment by miri64** [comment]
Our release process did not change, so yes!

**Comment by maribu** [comment]
Thanks for the reviews :-)
